docs/getting-started/quickstart-programming-python.md

```diff
@@ -0,0 +1,325 @@
+    return note
+```
+
+We add a new function called `sign_guestbook` that uses `requests` to send an HTTP POST request to the guestbook to record a greeting.
+We annotate it with [`@DBOS.communicator`](../tutorials-python/communicator-tutorial.md) to tell DBOS this function makes an external API call.
+As we'll see later, these annotations allow DBOS to provide lightweight durable execution.
+
+Stop your app with CTRL+C then restart it with `dbos start`. Make a few visits to the greeting URL in your browser (http://localhost:8000/greeting/Mike). With every new visit, the app should now print first that it has recorded your greeting in the guestbook, then that it has recorded your greeting in the database.
+
+```
+14:54:13 [    INFO] (dbos:main.py:39) >>> STEP 1: Signed the Guestbook: {"ip_address": "...", "greeted_name": "Mike", "greeted_ts": "2024-08-29T21:54:13.832Z"}
+14:54:13 [    INFO] (dbos:main.py:47) >>> STEP 2: Greeting to Mike recorded in the database!
+```
+
+You can visit the URL `https://demo-guestbook.cloud.dbos.dev/greetings/your-key-value` to see all the Guestbook greetings made with your key. Old greetings and keys are removed after a few days.
+
+## 5. Composing Reliable Workflows
+
+Next, we want to make our app reliable: guarantee that it inserts exactly one database record per guestbook signature, even if interrupted or restarted. This is called **durable execution**. DBOS makes this easy with [workflows](../tutorials/workflow-tutorial.md). To see them in action, change your `main.py` like so:
+
+```python
+import json
+import os
+
+import requests
+from dbos import DBOS
+from fastapi import FastAPI
+from fastapi.responses import HTMLResponse
+from schema import greetings
+
+app = FastAPI()
+DBOS(app)
+
+# Omitted for brevity: def readme()
+# Omitted for brevity: def get_greetings()
+
+@DBOS.communicator()
+def sign_guestbook(name: str):
+    key = os.environ.get("GUESTBOOK_KEY", None)
+    if key is None or len(key) != 36:
+        raise Exception("Please set the guestbook key in dbos-config.yaml")
+
+    url = "https://demo-guestbook.cloud.dbos.dev/record_greeting"
+    headers = {"Content-Type": "application/json"}
+    payload = {"key": key, "name": name}
+
+    response = requests.post(url, headers=headers, json=payload)
+    response_str = json.dumps(response.json())
+    if not response.ok:
+        raise Exception(f"Error signing guestbook: {response_str}")
+
+    DBOS.logger.info(f">>> STEP 1: Signed the Guestbook: {response_str}")
+
+
+@DBOS.transaction()
+def insert_greeting(name: str, note: str):
+    DBOS.sql_session.execute(greetings.insert().values(name=name, note=note))
+    DBOS.logger.info(f">>> STEP 2: Greeting to {name} recorded in the database!")
+
+
+@DBOS.workflow()
+def greeting_workflow(name: str, note: str):
+    sign_guestbook(name)
+    for _ in range(5):
+        DBOS.logger.info("Press Control + C to stop the app...")
+        DBOS.sleep(1)
+    insert_greeting(name, note)
```

Comment:
I'd add the types hint everywhere.

Reply:
For some of these, the real types are very clumsy to write and would distract from the example.